libs/sdk-core/src/ldk/store/locking_store.rs

```diff
@@ -0,0 +1,213 @@
+        let versioned_tl = Mutex::new(VersionedTimeLock { tl, version });
+
+        let locking_store = Self {
+            inner: store,
+            versioned_tl,
+        };
+        locking_store.lock().await?;
+
+        Ok(locking_store)
+    }
+
+    /// Refreshes the distributed lock to extend its duration.
+    ///
+    /// This method should be called periodically to maintain ownership of the lock.
+    ///
+    /// # Returns
+    /// * `Ok(SystemTime)` - The time until which the lock should be refreshed
+    /// * `Err(Error::Conflict)` - If the lock was lost or cannot be refreshed
+    /// * `Err(Error::Internal)` - If there was an internal error
+    pub async fn refresh_lock(&self) -> Result<SystemTime, Error> {
+        // Note: An optimization is possible here.
+        // If we failed to lock here, we can still try to lock again,
+        // asserting that the previous lock was acquired by us
+        // (i.e., no one acquired/released the lock in between).
+        let locked_until = self.lock().await?;
+        Ok(locked_until - Self::REFRESH_WINDOW)
+    }
+
+    async fn lock(&self) -> Result<SystemTime, Error> {
+        // Hold the lock on `self.versioned_tl` during the whole call to avoid
+        // race conditions with `unlock()` method.
+        let mut versioned_tl = self.versioned_tl.lock().await;
+        let lock_data = versioned_tl.tl.next_lock();
+        let value = lock_data
+            .encode()
+            .map_err(|e| Error::Internal(format!("Failed to encode lock_data: {e}")))?;
+        self.inner
+            .put(Self::KEY.to_string(), value, versioned_tl.version)
+            .await?;
+        versioned_tl.version += 1;
+        versioned_tl.tl.update_lock(&lock_data);
+        Ok(lock_data.locked_until)
+    }
+
+    /// Releases the distributed lock.
+    ///
+    /// This method should be called when the instance is done using the store
+    /// to allow other instances to acquire the lock. If the lock has already
+    /// expired or been lost, this method will still succeed.
+    ///
+    /// # Returns
+    /// * `Ok(())` - If the lock was successfully released or was already expired
+    /// * `Err(Error::Internal)` - If there was an internal error during release
+    pub async fn unlock(&self) -> Result<(), Error> {
+        // Hold the lock on `self.versioned_tl` during the whole call to avoid
+        // race conditions with `lock()` method.
+        let mut versioned_tl = self.versioned_tl.lock().await;
+        if let Some(lock_data) = versioned_tl.tl.unlock() {
+            let value = lock_data
+                .encode()
+                .map_err(|e| Error::Internal(format!("Failed to encode lock_data: {e}")))?;
+            self.inner
+                .put(Self::KEY.to_string(), value, versioned_tl.version)
+                .await?
+        }
+        Ok(())
+    }
+
+    async fn ensure_locked(&self) -> Result<(), Error> {
+        if !self.versioned_tl.lock().await.tl.is_locked() {
```

Comment:
This is called for every get/put/delete/list. And `lock` will always `put` the lock on the remote storage. I think this can be much more time efficient if the lock is only refreshed when necessary, i.e. the time lock is about to expire (with the refresh window as leeway).

Also this now blocks parallel calls, because each call will fight for the same mutex to lock the database.

Reply:
Ah!